contributors/design-proposals/encryption.md

```diff
@@ -0,0 +1,436 @@
+  on ARM
+
+There's no problem with adding this as an alternative later.
+
+## Configuration
+
+We will add the following options to the API server. At API server startup the
+user will specify:
+
+```yaml
+--encryption-provider-config=/path/to/config
+--encryption-provider=default
+--encrypt-resource=v1/Secrets
+```
+
+The encryption provider will check it has the keys it needs and if not, generate
+them as described in the following section.
+
+## Key Generation, Distribution and Rotation
+
+To start with we want to support a simple user-driven key generation,
+distribution and rotation scheme. Automatic rotation may be achievable in the
+future.
+
+To enable key rotation a common pattern is to have keys used for resource
+encryption encrypted by another set of keys (Key Encryption Keys aka KEK).  The
+keys used for encrypting kubernetes resources (Data Encryption Keys, aka DEK)
+are generated by the apiserver and stored encrypted with one of the KEKs.
+
+In future versions, storing a KEK off-host and off-loading encryption/decryption
+of the DEK to AWS KMS, Google Cloud KMS, Hashicorp Vault etc. should be
+possible. The decrypted DEK would be cached locally after boot.
+
+Using a remote encrypt/decrypt API offered by an external store will be limited
+to encrypt/decrypt of keys, not the actual resources for performance reasons.
+
+Incremental deliverable options are presented below.
+
+### Option 1: Simple list of keys on disk
+
+In this solution there is no KEK/DEK scheme, just single keys in a list on disk.
+They will live in a file specified by the --encryption-provider-config, which
+can be an empty file when encryption is turned on.
+
+If the key file is empty or the user calls PUT on a /rotate API endpoint keys
+are generated as follows:
+
+1. A new encryption key is created.
+1. The key is added to the config file on the API master with metadata including
+   an ID and an expiry time. Subsequent calls to rotate will prepend new keys to
+   the file such that the first key is always the key to use for encryption.
+1. The list of keys being used by the master is updated in memory so that the
+   new key is in the list of read keys.
+1. The list of keys being used by the master is updated in memory so that the
+   new key is the current write key.
+1. All secrets are re-encrypted with the new key.
+
+Pros:
+
+ - Simplicity.
+ - The generate/write/read interfaces can be pluggable for later replacement
+   with external secret management systems.
+ - A single master shouldn't require API Server downtime for rotation.
+ - No unseal step on startup since the file is already present.
+ - Attacker with access to /rotate is a DoS at worst, it doesn't return any
+   keys.
+
+Cons:
+
+ - Coordination of keys between a deployment with multiple masters will require
+   shutting down all masters, and manually generating and providing a key in the
+   key file.
+ - Users will be responsible for backing up the keyfile from the API server
+   disk.
```

Comment:
You can avoid this con with a kvno (key version number) and allowing a master that does not have a new version key to call out to get it dynamically.
The only problem left is that until all masters get to the new key you may get new secrets stored with an old key.
Other software deal with this by preloading a new key, but not activating it for encryption until there is confirmation all masters have the new key, at that point all switch to use the new key, and after a short delay a batch job to rotate all keys is started.